node/src/components/block_synchronizer/tests.rs

```diff
@@ -336,3 +452,1231 @@ async fn should_not_stall_after_registering_new_era_validator_weights() {
         }) if peers.contains(&peer) && id.block_hash == *block.hash()
     );
 }
+
+#[test]
+fn duplicate_register_block_not_allowed_if_builder_is_not_failed() {
+    let mut rng = TestRng::new();
+    let test_env = TestEnv::random(&mut rng);
+    let block = test_env.block();
+    let validator_matrix = test_env.gen_validator_matrix();
+    let mut block_synchronizer = BlockSynchronizer::new_initialized(&mut rng, validator_matrix);
+
+    // Register block for forward sync
+    assert!(block_synchronizer.register_block_by_hash(*block.hash(), false, true));
+    assert!(block_synchronizer.forward.is_some()); // we only get global state on historical sync
+
+    // Registering the block again should not be allowed until the sync finishes
+    assert!(!block_synchronizer.register_block_by_hash(*block.hash(), false, true));
+
+    // Trying to register a different block should replace the old one
+    let new_block = Block::random(&mut rng);
+    assert!(block_synchronizer.register_block_by_hash(*new_block.hash(), false, true));
+    assert_eq!(
+        block_synchronizer.forward.unwrap().block_hash(),
+        *new_block.hash()
+    );
+}
+
+#[tokio::test]
+async fn historical_sync_gets_peers_form_both_connected_peers_and_accumulator() {
+    let mut rng = TestRng::new();
+    let mock_reactor = MockReactor::new();
+    let test_env = TestEnv::random(&mut rng);
+    let block = test_env.block();
+    let validator_matrix = test_env.gen_validator_matrix();
+    let mut block_synchronizer = BlockSynchronizer::new_initialized(&mut rng, validator_matrix);
+
+    // Register block for historical sync
+    assert!(block_synchronizer.register_block_by_hash(*block.hash(), true, true));
+    assert!(block_synchronizer.historical.is_some());
+
+    let effects = block_synchronizer.handle_event(
+        mock_reactor.effect_builder(),
+        &mut rng,
+        Event::Request(BlockSynchronizerRequest::NeedNext),
+    );
+    assert_eq!(effects.len(), 2);
+    let events = mock_reactor.process_effects(effects).await;
+
+    // The first thing the synchronizer should do is get peers.
+    // For the historical flow, the synchronizer will get a random sampling of the connected
+    // peers and also ask the accumulator to provide peers from which it has received information
+    // for the block that is being synchronized.
+    assert_matches!(
+        events[0],
+        MockReactorEvent::NetworkInfoRequest(NetworkInfoRequest::FullyConnectedPeers {
+            count,
+            ..
+        }) if count == MAX_SIMULTANEOUS_PEERS
+    );
+
+    assert_matches!(
+        events[1],
+        MockReactorEvent::BlockAccumulatorRequest(BlockAccumulatorRequest::GetPeersForBlock {
+            block_hash,
+            ..
+        }) if block_hash == *block.hash()
+    )
+}
+
+#[tokio::test]
+async fn fwd_sync_gets_peers_only_form_accumulator() {
```

Comment:
```suggestion
async fn fwd_sync_gets_peers_only_from_accumulator() {
```

Reply:
Fixed in https://github.com/casper-network/casper-node/pull/3816/commits/8c830a518622c78e4f57c74fa331707740e90648